### PR TITLE
perf(nano): defer vector deletes to flush time

### LINKED ISSUE #3679: [Bug]:[Performance] O(n²) ingestion slowdown: per-relation delete() copies the entire vector matrix (NanoVectorDB)
### Do you need to file an issue?

- [x] I have searched the existing issues and this bug is not already filed.
- [x] I believe this is a legitimate bug, not just a question or feature request.

### Describe the bug

**Ingestion gets slower with every document, and never recovers.** On a 778-document run it went from ~150 docs/hour in the first hour to ~10 docs/hour by document 500, while the LLM and embedding services sat idle and the server pinned one CPU core at 92%.

**Cause: merging one relation copies the whole vector matrix.**

```python
# lightrag/operate.py:3332, inside _merge_edges_then_upsert — runs once per relation
await relationships_vdb.delete([rel_vdb_id, rel_vdb_id_reverse])   # clear the stale rows
...
await relationships_vdb.upsert(vdb_data)                            # write the new row
```

`NanoVectorDBStorage.delete` removes the ids from the live index immediately (`nano_vector_db_impl.py:492`), and `NanoVectorDB.delete` does that with `np.delete(matrix, index, axis=0)` (`nano_vectordb/dbs.py:133`) — a rebuild of the entire matrix.

So the cost is **O(rows in the index) per relation**, not per document:

| document | relations | matrix copies | cost of each copy |
|---|---|---|---|
| #1 | 400 | 400 | small index |
| #500 | 400 | 400 | 110k rows / 2.4 GB |

Total work across an ingestion is quadratic in corpus size. That is the decay curve.

**Measured** (400-relation document, synthetic, script below):

| rows already in the index | merge phase | flush | total |
|---|---|---|---|
| 2,000 | 0.35s | 0.14s | **0.49s** |
| 10,000 | 5.49s | 0.68s | **6.17s** |
| 40,000 | 23.01s | 2.27s | **25.28s** |

The flush is flat. The merge phase is what explodes.

**`py-spy record -d 30` on the live server**, self time by function:

```
77.3%  delete (numpy/lib/_function_base_impl.py:5395)
 6.1%  delete (numpy/lib/_function_base_impl.py:5376)
 2.6%  delete (numpy/lib/_function_base_impl.py:5380)
 1.4%  delete (nano_vectordb/dbs.py:133)
 1.2%  delete (nano_vectordb/dbs.py:127)
```

`py-spy dump`, the live stack (line numbers omitted — the profile is from our
v1.5.7 deployment, while the code references in this issue are against
`main` @ b81bb92):

```
delete                      numpy/lib/_function_base_impl.py
delete                      nano_vectordb/dbs.py
delete                      lightrag/kg/nano_vector_db_impl.py
_merge_edges_then_upsert    lightrag/operate.py
_locked_process_edges       lightrag/operate.py
```

89% of process CPU is one numpy call.

Only the NanoVectorDB backend is affected. Qdrant and PostgreSQL already buffer deletes (`_pending_vector_deletes`) instead of applying them per call.

---

**Second, smaller problem from the same eager delete: a removal can be silently lost.**

`delete` drops the row from `self._client` while the change is still unsaved. `index_done_callback` then reloads unconditionally if another writer has committed (`nano_vector_db_impl.py:688`), and a reload *replaces* `self._client` with a fresh parse of the file. The unsaved removal is gone and the row is back. No error, no log line.

`finalize` guards its own reload with `if not self._client_dirty` and its comment describes exactly this hazard for pending upserts. `index_done_callback` has no such guard.

When this is reachable: a second writer for the same namespace must commit between this writer's `delete` and its flush. A single-process, single-writer run will not hit it, because a writer resets its own stale flag right after broadcasting. A multi-worker deployment does have the window — `WORKERS` is a supported setting (`env.example`) and `finalize_storages` runs from the server lifespan shutdown (`lightrag/api/lightrag_server.py:1492`), outside the pipeline busy gate, so a peer worker exiting during a rolling restart is enough.

### Steps to reproduce

Save next to a checkout of `main`, run `python repro.py`. No LLM or embedding service needed (the embedding function is a local stub). Reproduces both problems.

```python
import asyncio
import shutil
import tempfile
import time

import numpy as np

from lightrag.kg.nano_vector_db_impl import NanoVectorDBStorage
from lightrag.kg.shared_storage import finalize_share_data, initialize_share_data
from lightrag.utils import EmbeddingFunc

DIM = 1024


async def _embed(texts, **kwargs):
    return np.random.rand(len(texts), DIM).astype(np.float32)


def _storage(working_dir):
    return NanoVectorDBStorage(
        namespace="repro",
        workspace="ws",
        global_config={
            "working_dir": working_dir,
            "embedding_batch_num": 128,
            "vector_db_storage_cls_kwargs": {"cosine_better_than_threshold": 0.2},
        },
        embedding_func=EmbeddingFunc(embedding_dim=DIM, max_token_size=512, func=_embed),
        meta_fields={"content"},
    )


async def problem_1_cost_grows_with_index_size():
    """One document's merge: delete each relation's stale rows, then upsert it."""
    print("Problem 1 - merge cost of one 400-relation document")
    print(f"  {'rows already indexed':>22}  {'merge':>9}  {'flush':>8}  {'total':>9}")
    for existing in (2_000, 10_000, 40_000):
        working_dir = tempfile.mkdtemp()
        try:
            initialize_share_data()
            storage = _storage(working_dir)
            await storage.initialize()
            await storage.upsert({f"e{i}": {"content": f"c{i}"} for i in range(existing)})
            await storage.index_done_callback()

            started = time.perf_counter()
            for i in range(400):
                await storage.delete([f"e{i}", f"rev-e{i}"])
                await storage.upsert({f"e{i}": {"content": f"updated{i}"}})
            merge = time.perf_counter() - started

            started = time.perf_counter()
            await storage.index_done_callback()
            flush = time.perf_counter() - started

            print(f"  {existing:>22,}  {merge:>8.2f}s  {flush:>7.2f}s  {merge + flush:>8.2f}s")
            finalize_share_data()
        finally:
            shutil.rmtree(working_dir, ignore_errors=True)


async def problem_2_delete_lost_across_writers():
    """A delete is dropped if another writer commits before this writer flushes."""
    print("\nProblem 2 - delete lost across writers")
    working_dir = tempfile.mkdtemp()
    try:
        initialize_share_data()
        writer = _storage(working_dir)
        other = _storage(working_dir)
        await writer.initialize()
        await other.initialize()

        await writer.upsert({"a": {"content": "a"}, "b": {"content": "b"}})
        await writer.index_done_callback()

        await writer.delete(["a"])
        await other.upsert({"c": {"content": "c"}})
        await other.index_done_callback()      # flags `writer` as stale
        await writer.index_done_callback()     # reloads, dropping the unsaved removal

        reader = _storage(working_dir)
        await reader.initialize()
        ids = sorted(row["__id__"] for row in (await reader.client_storage)["data"])
        print(f"  on disk: {ids}")
        print(f"  expected: ['b', 'c']")
        print(f"  'a' resurrected: {'a' in ids}")
        finalize_share_data()
    finally:
        shutil.rmtree(working_dir, ignore_errors=True)


async def main():
    await problem_1_cost_grows_with_index_size()
    await problem_2_delete_lost_across_writers()


asyncio.run(main())
```

Output on `main` @ b81bb92, stable across repeated runs:

```
Problem 1 - merge cost of one 400-relation document
    rows already indexed      merge     flush      total
                   2,000      0.35s     0.14s      0.49s
                  10,000      5.49s     0.68s      6.17s
                  40,000     23.01s     2.27s     25.28s

Problem 2 - delete lost across writers
  on disk: ['a', 'b', 'c']
  expected: ['b', 'c']
  'a' resurrected: True
```

### Expected Behavior

Merging one document should cost the same whether the index holds 2k rows or 40k. Removing 800 ids should not require 400 separate full-matrix rebuilds.

`delete` followed by `index_done_callback` should leave the row deleted on disk, whether or not another writer committed in between.

### LightRAG Config Used

The reproduction needs no configuration. Where we hit it:

```
LIGHTRAG_VECTOR_STORAGE=NanoVectorDBStorage   (default)
LIGHTRAG_GRAPH_STORAGE=NetworkXStorage        (default)
LIGHTRAG_KV_STORAGE=JsonKVStorage             (default)
MAX_PARALLEL_INSERT=8
EMBEDDING_DIM=4096
```

778 documents, mostly data-entry rule documents yielding ~200 entities and ~400 relations each. `vdb_relationships.json` reached 2.4 GB / ~110k rows.

### Logs and screenshots

Included inline above (py-spy profile and stack).

### Additional Information

- LightRAG Version: 1.5.7 (also 1.5.6)
- Operating System: Ubuntu, Linux 5.4
- Python Version: 3.12
- Related Issues: #1957 reports the same decay curve from the merge stage without identifying this cause. #3145 added the deferred-embedding buffer that a fix can mirror.

### PR DESCRIPTION
**Pull request by 0xPabloxx**
## Summary

Ingestion on the NanoVectorDB backend slows down quadratically because merging each relation rebuilds the whole vector matrix. This queues the deletes and applies them once per flush instead of once per relation.

On a synthetic 400-relation merge over a 40k-row index: **25.28s → 2.50s**, and the merge phase itself **23.01s → 0.02s**.

Full symptom, profile and reproduction: #3679 .

## Why

`_merge_edges_then_upsert` clears a relation's stale forward/reverse rows before writing the new one, once per relation (`operate.py:3332`, same pattern at `operate.py:2270`):

```python
await relationships_vdb.delete([rel_vdb_id, rel_vdb_id_reverse])
await relationships_vdb.upsert(vdb_data)
```

`NanoVectorDBStorage.delete` applies that to the live index immediately (`nano_vector_db_impl.py:492`), and `NanoVectorDB.delete` implements removal as `np.delete(matrix, index, axis=0)` (`nano_vectordb/dbs.py:133`), which rebuilds the entire matrix. Cost is O(rows in the index) per call, so a 400-relation document performs 400 full-matrix copies, and every document costs more than the last.

On a 778-document run (`vdb_relationships.json` at 2.4 GB / ~110k rows) `py-spy record -d 30` attributed 89% of process CPU to `np.delete`, with the LLM and embedding services idle. Throughput fell from ~150 docs/hour to ~10.

The same eager removal can also lose a delete. The row leaves `self._client` while the change is unsaved, and `index_done_callback` reloads unconditionally when another writer has committed (`nano_vector_db_impl.py:688`); a reload replaces `self._client`, so the row comes back with no error. `finalize` guards its own reload with `if not self._client_dirty` and its comment describes this hazard; `index_done_callback` does not. This needs a second writer committing inside the window, so a single-process run will not hit it — a multi-worker deployment can, since `finalize_storages` runs from the lifespan shutdown (`lightrag/api/lightrag_server.py:1492`) outside the pipeline busy gate.

## What changed

One file plus tests: `nano_vector_db_impl.py` (+111/−28, most of it docstring). **No changes to `operate.py`** — the merge hot path is untouched.

`delete` cancels the matching pending upsert and queues the id in `_pending_deletes` instead of touching `self._client`. `_flush_pending_locked` applies every queued id in one `client.delete` call, **before** materializing pending upserts, so an id deleted and re-upserted in the same batch ends up with exactly the new row.

This mirrors the deferred-embedding protocol from #3145 and the delete buffers already used by the Qdrant and PostgreSQL backends.

This narrows the delete-loss window; the redo log added in dfc865b closes it for deletes.

Buffer bookkeeping follows the sibling backends:

- The read paths report a queued id as absent.
- An `upsert` for the same id cancels the queued delete — `client.upsert` overwrites the row in place, so applying the delete first would be wasted work.
- `query` and `client_storage` keep reading the materialized index. Filtering queued ids out of query results *after* top-k selection would silently shrink the result set, since the removed rows occupy slots that never get refilled. Qdrant documents the same contract at `qdrant_impl.py:756`.
- `drop` and `drop_pending_index_ops` discard the queue alongside pending upserts.
- `delete_entity` / `delete_entity_relation` prune ids their eager delete already covered.
- `finalize` no longer rewrites the whole file when a flush changed nothing (which also broadcast a reload to every other process).

## Side effect analysis

- **Default values**: unchanged, no new configuration.
- **Backward compatibility**: no API or storage-format change. `delete` keeps its signature and its "in-memory only until `index_done_callback`" contract; only the moment the removal reaches `self._client` moves.
- **Downstream consumers**: checked every `.delete(` caller on a vector store. The merge paths (`operate.py:2270`, `3332`) and manual graph edits (`utils_graph.py:1136` and three others) depend on the state after the flush, preserved by the delete-before-upsert ordering. The purge path (`lightrag.py:5485`, `5536`, `5556`) uses a plain `_insert_done`, not `_discard_pending_index_ops`, so queued deletes are not dropped there.
- **Query semantics**: a queued delete stays visible to `query` until the flush, where the row was previously already gone. This matches the documented behaviour of the other buffering backends and is bounded by the same flush that already gates pending upserts.
- **Not converted**: `delete_entity` / `delete_entity_relation` still delete eagerly. They are off the merge hot path and the class docstring already marks them as non-pipeline entry points, so this PR only keeps them consistent with the queue. Happy to fold them in if you would rather have one path.
- **Considered and rejected**: fixing the lost delete on its own by adding `if not self._client_dirty` to `index_done_callback`'s reload, mirroring `finalize`. That trades one data loss for another — skipping the reload discards the other writer's committed rows, and the following save overwrites them.

## Validation

The reproduction from #3679 , on this branch:

```
Problem 1 - merge cost of one 400-relation document
    rows already indexed      merge     flush      total
                   2,000      0.02s     0.14s      0.16s
                  10,000      0.02s     0.57s      0.59s
                  40,000      0.02s     2.48s      2.50s

Problem 2 - delete lost across writers
  on disk: ['b', 'c']
  'a' resurrected: False
```

On `main` the same rows are 0.49s / 6.17s / 25.28s with `'a' resurrected: True`. The merge phase is now flat at 0.02s: the matrix is rebuilt once per flush instead of once per relation. What remains is the flush, which was always there.

**Tests**: `pytest tests/kg/nano_impl` — 30 passed. Two tests in `test_nano_deferred_embedding.py` pinned the eager-delete behaviour and are updated in the first commit, so the suite is green at each commit.

`test_nano_deferred_delete.py` covers batching (one `client.delete` per flush, not one per call), delete-before-upsert ordering, the read-your-writes paths, `query` not truncating top-k, upsert superseding a queued delete, `drop`, delete-only flushes, `finalize` skipping a no-op save, and crash rollback. `test_delete_survives_a_concurrent_commit_by_another_writer` fails on the eager-delete behaviour and passes here.

Also green: `tests/kg/json_impl` (74), `tests/kg/networkx_impl` (81), `tests/kg/noop_impl` (6), `tests/setup` (277). I could not run the full offline suite locally because several optional backends are not installed (`opensearchpy`, `jwt`), so those directories error at collection on `main` as well; nothing here touches them.

**Lint**: `ruff check --ignore=E402` and `ruff format --check` clean, with ruff pinned at 0.15.17 per `.pre-commit-config.yaml`.

## Related Issues

Fixes #3679 
Related to #1957 — the decay reported there matches this signature on the Nano backend, though that issue also covers backends this change does not touch.

## Checklist

- [x] Changes tested locally
- [x] Code reviewed
- [x] Documentation updated (if necessary) — the class docstring gains a "Deferred-delete protocol" section beside the existing deferred-embedding one
- [x] Unit tests added (if applicable)

### CODE REVIEW

**Comment by danielaskdd** [comment]
@codex review

**Comment by danielaskdd** [comment]
Thanks — the profile, the reproduction and the design write-up are all excellent, and the direction is right: deferring the delete and applying it in one batched call per flush is what the five server-backed backends already do, so this brings Nano in line rather than inventing anything. The `operate.py` hot path staying untouched is the right call too.

I want the Codex P2 on `_flush_pending_locked` addressed before merge. I traced it against your branch and it holds up:

1. `_flush_pending_locked` applies the queued deletes to `self._client`, sets `_client_dirty = True`, and clears the queue at line 383 — **before** the save.
2. `_save_to_disk_locked()` (line 759) raises. `_client_dirty` stays `True`; `_pending_deletes` is already empty, so there is no replay record left.
3. Another process commits → our `storage_updated.value` flips.
4. The next `index_done_callback` opens with an **unconditional** `_reload_client_from_disk_locked(for_write=True)` (line 752), which does `self._client = NanoVectorDB(...)` — a wholesale replacement. The deleted row is back, nothing reapplies it, and the following save succeeds. Silent loss.

So the PR narrows the delete-loss window from `[delete call → next successful save]` down to `[flush-applied-with-failed-save → next reload]`, rather than closing it — the claim "the lost-delete problem goes away as a consequence" is a bit stronger than what the change delivers. To be clear, this is **not a regression you introduced**: that residual window is exactly the one upserts already have on `main` (the flush clears `_pending_upserts` at line 454, same unconditional reload), so you are inheriting a pre-existing asymmetry, not creating one. I have filed #3688 for the general case (Nano upserts + both halves on Faiss).

But the delete half belongs in this PR, because delete durability semantics are precisely what it is about, and Codex's first suggestion is the structurally right shape rather than a patch:

**Requested change** — retain `_pending_deletes` until the save is durable.

- Drop the `self._pending_deletes.clear()` at line 383; clear the queue **after** `_save_to_disk_locked()` succeeds, in both `index_done_callback` and `finalize`.
- That makes the reload path lossless in both directions: the reload picks up the other writer's committed rows, and the retained queue replays ours on top. Replay is idempotent — a replayed id that is no longer present simply matches nothing.
- Please keep your existing rejection of the alternative (guarding `index_done_callback`'s reload on `_client_dirty`) — I agree with your reasoning there, and it is worth noting *why* retention is strictly better: in the residual state, `finalize`'s guarded path is not lossless either, it just loses the *other* writer's rows instead of ours. Retention is the only shape that loses neither.

Two details to decide explicitly while you are in there:

- **A flush whose deletes matched nothing has nothing to persist**, so it can clear the queue immediately (`deleted_count == 0`) — only matched-but-unsaved ids need retention. Otherwise a queue full of ids that were never in the index stays resident indefinitely, and `finalize`'s `if not self._client_dirty: return` early exit (line 992) would keep skipping past it.
- Keep `_client_dirty` derived strictly from "did the materialized store actually change", as you have it now — a replayed no-op delete must not flag the storage dirty.

**Test**: one fix-proof case that fails on the current branch behaviour — queue a delete, make the save raise, have a second writer commit, run `index_done_callback` again, then assert the row is gone from the on-disk snapshot *and* the other writer's rows survived. `test_delete_survives_a_concurrent_commit_by_another_writer` covers the pre-flush window; this is the post-flush one.

Everything else in the PR I am happy with as-is: the delete-before-upsert ordering, `query`/`client_storage` keeping the materialized view (not filtering after top-k — agreed, and thanks for citing the Qdrant contract), leaving `delete_entity` / `delete_entity_relation` eager with the redundant-queue pruning, and the `finalize` no-op-save skip. No need to fold the entity paths in.

For context: I have also opened #3681 to port this same protocol to Faiss, which is the last backend still deleting eagerly — with a note that the port should pick up whatever we settle on here.

**Review by dianatofficial — COMMENTED** [review]
Great catch on this edge case. The fallback logic looks robust.

**Comment by 0xPabloxx** [comment]
Thanks for tracing it that carefully — the walk-through matches what I get on the branch, and you are right that "the lost-delete problem goes away as a consequence" overstated it. I have corrected that claim in the protocol docstring and I will fix the PR description too.

Pushed `dfc865b`.

### What changed

Deletes now use two buffers rather than one, because the flush and the save can fail independently:

- `_pending_deletes` — queued, not yet applied to `self._client`. Unchanged.
- `_unsaved_deletes` — applied to `self._client` but not yet on disk, held as `id -> the row's __created_at__`. A redo log, cleared only after `_save_to_disk_locked()` returns, in both `index_done_callback` and `finalize`.

A flush replays the log against the current client, so the reload that used to resurrect a row now removes it again.

I split the two instead of keeping one set because the single set was doing double duty — before the flush it is a pending buffer, after the flush it is a redo log — and the two have opposite clearing rules. My first attempt kept one set and gated the clear on the pre-delete `_client_dirty`; it worked for your scenario but broke on a second consecutive save failure, where the replay matches nothing (the row is already gone from the client), and the ids were dropped while still unsaved.

### On the two details you asked me to decide

**Which ids to retain.** Only ids that actually removed a row enter the log. I ended up implementing this by computing the matched set before deleting rather than by testing `deleted_count == 0`, because `deleted_count` is a batch aggregate: one hit in a queue of `{matched, ghost1, ghost2}` would retain all three. That matters here — the merge stage deletes `rel_vdb_id_reverse` alongside `rel_vdb_id`, and the reverse id usually does not exist, so mixed batches are the norm rather than the edge case. Retaining a ghost id is not just residency: if another writer later publishes a row under that id, the replay deletes it. Test: `test_an_id_that_matched_nothing_is_never_replayed`.

Keeping only matched ids also fixes something I had regressed. With the whole queue retained, every failed-save retry re-issued `client.delete` over the full queue, so a run of failures rebuilt the matrix once per retry — exactly the cost this PR removes. Test: `test_repeated_failed_saves_do_not_re_delete_each_time` asserts one `client.delete` across four failed saves.

**`_client_dirty`.** Still derived strictly from whether the materialized store changed: it is set only inside `if matched:`. A replay that matches nothing does not flag the storage dirty.

### One hazard retention introduces, and how far I took it

Retention converts "lose a delete" into a different exposure: after a failed save, another writer can publish a row under an id we removed, and a naive replay would delete it. Ids are content hashes, so two workers touching the same entity produce the same id — this is reachable in the same multi-worker deployments that make the original window reachable.

I reproduced it, then made the replay match on `__created_at__` instead of on the id alone: a replayed id only removes the row it removed before, and a row written afterwards survives. Test: `test_replay_does_not_remove_a_newer_row_under_the_same_id`.

The limit is worth stating plainly: `upsert` stamps `__created_at__` with `int(time.time())`, so two writes inside the same second are indistinguishable and the guard cannot separate them. In practice the two writes are seconds apart and it holds — the test needs an explicit `sleep(1.1)` to leave the same-second window. If you want a hard guarantee here I think it needs a real per-row version rather than a timestamp, which is a wider change than this PR; happy to do it in a follow-up if you would rather have it.

### Documentation

Two places were left inaccurate by the change and are corrected:

- `drop_pending_index_ops` says materialized-but-unsaved records are deliberately not rolled back, and justifies it by reprocessing re-issuing the same deterministic ids. That argument is weaker for deletes: dropping the redo log does discard a removal that reached the client, and unlike an upsert a delete is only re-issued if the reprocess reaches the same merge. Noted in the docstring.
- The protocol section no longer claims the reload is lossless both ways. It holds for deletes; upserts have no redo log — `_pending_upserts` is cleared once the rows materialize — so a materialized-but-unsaved upsert is still dropped by the same reload. Pointed at #3688.

### Validation

`pytest tests/kg/nano_impl` — 35 passed.

`test_an_id_that_matched_nothing_is_never_replayed` fails on the previous commit of this branch. The other two guard hazards that retention itself introduces, so they pass there trivially — there is no replay to go wrong. I checked them against a retain-the-whole-queue-and-delete-by-id implementation instead, the shape I would have shipped without the two decisions above; all three fail on it.

Also re-ran the probes from the PR description on this commit: the merge-phase numbers are unchanged (0.02s flat at 2k/10k/40k rows), and the cross-writer delete still survives.

Also green, unchanged from the earlier run: `tests/kg/json_impl` (74), `tests/kg/networkx_impl` (81), `tests/kg/noop_impl` (6), `tests/setup` (277). `ruff check --ignore=E402` and `ruff format --check` clean at 0.15.17.

Thanks also for opening #3688 and #3681 — happy to help with the Faiss port once the shape here settles.

**Comment by danielaskdd** [comment]
Pushed `c99c63d` onto your branch — two gaps left by `dfc865b`, both the same defect the redo log closes, on paths the log does not yet reach. Everything else in `dfc865b` I took as-is.

First, on your side of it: the `__created_at__` guard is a catch I did not anticipate when I asked for retention. Retention does convert "lose a delete" into "delete a row we did not mean to", and matching the row rather than the id is the right shape for it. Same for computing the matched set instead of testing `deleted_count == 0` — you are right that mixed batches are the norm here, since the merge stage issues `rel_vdb_id_reverse` alongside `rel_vdb_id` and the reverse id usually does not exist. And thank you for stating the whole-second limit plainly rather than letting the test's `sleep(1.1)` imply a guarantee. A real per-row version is a follow-up, agreed — not this PR.

### 1. The eager entity paths are applied-and-unsaved too

`delete_entity` / `delete_entity_relation` staying eager is right — they are off the merge hot path, and I said as much. But their removals reach `self._client` while unsaved exactly like a flushed one, and they were not recorded, so the same failed save plus a foreign commit resurrects the row on the next reload with nothing left to reapply it. Both paths already hold the rows they remove, so recording `id -> __created_at__` there is three lines and matches your replay guard.

### 2. `drop_pending_index_ops` cleared the log

This one contradicts the rule stated one paragraph above it in its own docstring: records already materialized into `self._client` are deliberately not rolled back. Clearing the log is a half rollback — it neither restores the row nor keeps it gone, it only makes the removal fragile again, which is the state the log exists to end. It now discards `_pending_deletes` (buffered work an aborted batch gives up on) and keeps `_unsaved_deletes`. I also wrote in the rationale your comment described: the docstring note itself had not made it into the commit, the two `.clear()` lines were the whole change there.

### Tests

Four cases; `tests/kg/nano_impl` 39 passed.

`test_eager_entity_delete_is_replayed_across_a_reload` and `test_aborting_batch_keeps_the_redo_log` fail behaviourally on `dfc865b` — the row is back in the on-disk snapshot. The other two pin contracts `dfc865b` introduced without coverage: reads reporting a logged id as absent after a reload, and `upsert` cancelling the log entry. That second one is the case `__created_at__` cannot decide — a rewrite inside the same second as the deleted row carries the same stamp — so the cancel at `upsert` time is what makes it deterministic instead of timing-dependent. Worth a test standing on it before someone reads the guard as sufficient on its own.

### Two things I left alone

Notes rather than requests:

- The read paths match on id, not on `__created_at__`, so between a failed save and the replay a *newer* row another writer published under the same id reads as absent from this process. Narrow, and it errs toward hiding rather than serving deleted data, so I would rather document it in the protocol section than grow the read path.
- `_flush_pending_locked` reaches for `self._client._NanoVectorDB__storage` while the other two sites use `getattr(self._client, "_NanoVectorDB__storage")`. Works either way; pick one if you touch it.

From my side this is ready to merge once CI is green. For #3681, the Faiss port should pick up all of this: two buffers, match on the row rather than the id, log the eager paths, and keep the log across an abort.

**Comment by danielaskdd** [comment]
@codex review

**Comment by danielaskdd** [comment]
Pushed `ced8856` — this takes the collision-proof version now rather than in a follow-up. Codex came back on `c99c63d` with two P2s, and they turn out to be one root cause, which is what changed my mind about deferring.

### They are the same finding

`__created_at__` cannot identify a row, so the guard needed a second mechanism — cancelling the redo entry in `upsert` — and that mechanism is itself lossy:

1. `delete('x')` flushes, the save fails → redo entry `x -> T1`, disk still holds x@T1.
2. `upsert('x')` cancels the entry. The replacement row is buffered, not materialized.
3. The batch aborts → `drop_pending_index_ops` drops the buffered row.
4. Another writer commits → the reload restores x@T1, with nothing left to replay.

The cancel discards the only replay record while the replacement does not exist yet. Moving it to materialization time closes that but keeps the same-second collision; fixing the identity closes both and lets the cancel go away entirely.

### What changed

`_unsaved_deletes` is now `id -> _row_fingerprint(removed row)` — an md5 over `json.dumps(row, sort_keys=True)` of the whole stored record: id, timestamp, content, base64 vector, every meta field. Stable across a save/reload round-trip — sorted keys make key order irrelevant, tuples and lists both render as arrays, and `__vector__` is stripped into the matrix before the record is stored. A replay removes only a row that still matches, so anything written under that id since — by another writer or by us — survives.

`upsert` no longer touches the redo log, only the queued-delete set. Your concurrency test drops its `sleep(1.1)` and now exercises the same-second case it used to step around; it fails on `c99c63d` with the newer row destroyed.

`delete_entity` / `delete_entity_relation` record fingerprints too.

### The one case a fingerprint cannot separate

A rewrite byte-identical to the removed row. Materializing such a row retires the entry, at the bottom of the flush — and that is exactly the case where retiring it costs nothing, since a resurrection would restore the row that is present anyway. Only rows written by that upsert can match there: a resurrected one was already removed by the replay at the top of the same flush.

What remains is a byte-identical row published by *another* writer, which no content-based identity can tell from the row we deleted. That is documented in the protocol section rather than fixed — a per-row version counter is a storage-format change, and I do not think this residue earns one.

### Cost

One md5 per matched row per flush, computed only for ids in the two buffers, so the replay costs nothing while the log is empty — the normal path. 800 fingerprints (a 400-relation document, forward plus reverse) over 1536-dim rows is 22 ms, against a merge phase this PR takes from 23.01s to 0.02s. The 40k-row probe is unchanged on the same machine: merge 0.079s / flush 0.225s here, 0.081s / 0.216s on `c99c63d`.

### Tests

`tests/kg/nano_impl` — 41 passed, and 1.1s quicker without the sleep.

Two fail behaviourally on `c99c63d`: `test_replay_does_not_remove_a_newer_row_under_the_same_id` (newer row destroyed) and `test_abort_after_a_rewrite_keeps_the_redo_entry` (old row resurrected). `test_a_rewrite_identical_to_the_removed_row_drops_the_redo_entry` pins the retirement rule with a frozen clock, since that identity needs both the same content and the same whole second.

I also folded in the `getattr(self._client, "_NanoVectorDB__storage")` alignment while rewriting those lines, so the earlier style note is resolved.

For #3681, the Faiss port inherits this shape: two buffers, fingerprint identity, no cancel on upsert, and keep the log across an abort.

**Comment by danielaskdd** [comment]
@codex review

**Comment by danielaskdd** [comment]
Pushed `57da716` for the P1. Confirmed the example exactly: after a failed delete of `id1` and a foreign commit of `id3`, `finalize()` leaves the file holding `id2` alone.

**The finding is right and the reasoning is the one this PR unlocked.** `finalize` skips the stale-client reload whenever `self._client` is dirty, on the grounds that the reload would drop materialized-but-unsaved rows. That holds for upserts. It stopped holding for removals the moment the redo log landed: reload *and* replay loses neither side, while skipping loses the other writer's rows.

**Not a regression, though.** On `main` the same overwrite is reachable without any save failing — `delete()` there applies eagerly and dirties the client, so `finalize` walks straight past the reload into the save. What is new here is only that it became fixable.

**The fix** is `_unsaved_upserts`: a flag for the half of the dirty state that a reload really would destroy. Set only where `client.upsert` materializes rows, cleared wherever `_client_dirty` is, and it — not `_client_dirty` — now gates the reload in `finalize`. `_client_dirty` keeps its own job of deciding whether a save is owed. Removals set `_client_dirty` without setting the flag, so a delete-only retry reloads, replays the log on top, and saves both sides.

I kept it to `finalize`. `index_done_callback` still reloads unconditionally, so a materialized-but-unsaved upsert is still dropped there — that is #3688, and it wants a redo log for upserts rather than a guard in the reload path. Adding the guard there instead would just swap whose rows get lost, which is the trade we already rejected earlier in this PR.

**Tests** — `tests/kg/nano_impl` 43 passed.

`test_finalize_reloads_before_retrying_a_delete_only_save` is Codex's scenario and fails on `ced8856` with `id3` gone. `test_finalize_still_protects_unsaved_upserts_from_the_reload` pins the other half, so the guard cannot be loosened into dropping unsaved rows.

Also green: `tests/kg/json_impl`, `tests/kg/networkx_impl`, `tests/kg/test_file_atomic.py`, `tests/kg/test_unified_lock_safety.py` — 209 in total with `nano_impl`. `ruff check` and `ruff format --check` clean.

**Comment by danielaskdd** [comment]
@codex review

File: lightrag/kg/nano_vector_db_impl.py [hunk 1]
```diff
@@ -322,9 +441,47 @@ async def _flush_pending_locked(self) -> None:
...
+            queued = len(self._pending_deletes)
+            storage = getattr(self._client, "_NanoVectorDB__storage")
+            # A queued id removes whatever row is there; a replayed one only
+            # removes the row version it removed before, so anything written
+            # under that id since — by another writer or by us — survives.
+            matched: dict[str, str] = {}
+            for dp in storage["data"]:
+                doc_id = dp["__id__"]
+                if doc_id in self._pending_deletes:
+                    matched[doc_id] = self._row_fingerprint(dp)
```
[the remark this answers is not in the document]
Comment (danielaskdd) [reply lightrag/kg/nano_vector_db_impl.py:483]: Declining this one, and pinning the semantics so it does not get quietly changed later.

The two buffers are scoped differently on purpose. `_pending_deletes` holds a **request**: the caller named an id, so the flush removes whatever row carries that id — the contract the eager `delete()` had, and the one every server-backed backend implements (`DELETE WHERE id = ...`). `_unsaved_deletes` holds a **record** of a removal that already happened, so it has to be version-scoped: replaying it must not remove a row that has since taken the id's place. The asymmetry you spotted is that distinction, not an unfinished rollout of the fingerprint.

Version-scoping the request would trade a rare race for a systematic failure in the other direction. Purge (`adelete_by_doc_id`, `_purge_kg_contributions`) would silently skip vectors it was told to remove whenever another writer had touched the row, leaving orphans that only `lightrag-rebuild-vdb` can clean up. And "capture the target row version when establishing the delete intent" means an `O(rows)` lookup inside `delete()`, once per merged relation — precisely the per-call cost this PR exists to remove.

The window also needs two writers to upsert and delete the same id concurrently, which the single-writer invariant in the class docstring already excludes. That is what separates it from the redo-log hazard: a *replay* fires later, across reloads, and can span `finalize` at shutdown, which runs outside the pipeline gate — and guarding that costs nothing.

Both scopes are now stated on `delete` and in the protocol section, and `test_a_queued_delete_is_scoped_to_the_id_not_the_row` pins the behaviour. `419732a`.

File: lightrag/kg/nano_vector_db_impl.py [hunk 2]
```diff
@@ -716,11 +911,14 @@ async def get_by_id(self, id: str) -> dict[str, Any] | None:
...
         """
-        # Read-your-writes: a buffered upsert is visible before its flush.
+        # Read-your-writes: a buffered upsert is visible before its flush;
+        # an id queued for deferred delete (and not re-upserted) reads as absent.
         async with self._storage_lock:
             pending = self._pending_upserts.get(id)
             if pending is not None:
                 return self._format_record(pending.record)
+            if id in self._pending_deletes or id in self._unsaved_deletes:
+                return None
```
[the remark this answers is not in the document]
Comment (danielaskdd) [reply lightrag/kg/nano_vector_db_impl.py]: Agreed, fixed in `419732a`.

This is the divergence the fingerprint change opened: the replay goes out of its way to preserve a row another writer wrote over one we removed, and then the read paths suppressed it on id membership alone. The row stayed invisible until some later save cleared the log — in a failing-save loop, not a short window.

All three paths now look the row up and suppress it only when its fingerprint matches the entry (`_matches_redo_entry`, snapshotting the logged value under the lock and comparing after the client read). Queued ids keep the id-scoped suppression, so reads mirror the replay rules exactly: `_pending_deletes` by id, `_unsaved_deletes` by row version. Cost is a dict lookup while the log is empty, which is the normal path.

`test_reads_expose_a_row_written_over_a_logged_delete` fails on `57da716`; `test_reads_still_hide_the_row_a_redo_entry_names` pins the half that must stay hidden, so the guard cannot be loosened into serving a row we did delete.

**Comment by danielaskdd** [comment]
@codex review
